router/data_store.go

```diff
@@ -39,196 +33,322 @@ type SyncHandler interface {
 	Remove(id string) error
 }
 
-func NewEtcdDataStore(etcd EtcdClient, prefix string) DataStore {
-	return &etcdDataStore{
-		prefix:   prefix,
-		etcd:     etcd,
-		stopSync: make(chan bool),
-	}
-}
+type pgDataStore struct {
+	pgx *pgx.ConnPool
+
+	routeType string
+	tableName string
 
-type etcdDataStore struct {
-	prefix   string
-	etcd     EtcdClient
-	stopSync chan bool
+	doneo sync.Once
+	donec chan struct{}
 }
 
-var ErrExists = errors.New("router: route already exists")
-var ErrNotFound = errors.New("router: route not found")
+const (
+	routeTypeHTTP = "http"
+	routeTypeTCP  = "tcp"
+	tableNameHTTP = "http_routes"
+	tableNameTCP  = "tcp_routes"
+)
 
-func (s *etcdDataStore) Add(r *router.Route) error {
-	data, err := json.Marshal(r)
-	if err != nil {
-		return err
+// NewPostgresDataStore returns a DataStore that stores route information in a
+// Postgres database. It uses pg_notify and a listener connection to watch for
+// route changes.
+func NewPostgresDataStore(routeType string, pgx *pgx.ConnPool) *pgDataStore {
+	tableName := ""
+	switch routeType {
+	case routeTypeHTTP:
+		tableName = tableNameHTTP
+	case routeTypeTCP:
+		tableName = tableNameTCP
+	default:
+		panic(fmt.Sprintf("unknown routeType: %q", routeType))
 	}
-	_, err = s.etcd.Create(s.path(r.ID), string(data), 0)
-	if e, ok := err.(*etcd.EtcdError); ok && e.ErrorCode == 105 {
-		err = ErrExists
+	return &pgDataStore{
+		pgx:       pgx,
+		routeType: routeType,
+		tableName: tableName,
+		donec:     make(chan struct{}),
 	}
-	return err
 }
 
-func (s *etcdDataStore) Set(r *router.Route) error {
-	data, err := json.Marshal(r)
-	if err != nil {
-		return err
+const sqlAddRouteHTTP = `
+INSERT INTO ` + tableNameHTTP + ` (parent_ref, service, domain, tls_cert, tls_key, sticky)
+	VALUES ($1, $2, $3, $4, $5, $6)
+	RETURNING id, created_at, updated_at`
+
+const sqlAddRouteTCP = `
+INSERT INTO ` + tableNameTCP + ` (parent_ref, service, port)
+	VALUES ($1, $2, $3)
+	RETURNING id, created_at, updated_at`
+
+func (d *pgDataStore) Add(r *router.Route) (err error) {
+	switch d.tableName {
+	case tableNameHTTP:
+		err = d.pgx.QueryRow(
+			sqlAddRouteHTTP,
+			r.ParentRef,
+			r.Service,
+			r.Domain,
+			r.TLSCert,
+			r.TLSKey,
+			r.Sticky,
+		).Scan(&r.ID, &r.CreatedAt, &r.UpdatedAt)
+	case tableNameTCP:
+		err = d.pgx.QueryRow(
+			sqlAddRouteTCP,
+			r.ParentRef,
+			r.Service,
+			r.Port,
+		).Scan(&r.ID, &r.CreatedAt, &r.UpdatedAt)
```

Comment:
In the controller we strip the dashes from UUIDs before they hit the presentation layer. I can't remember the rationale for it, and I'm open to reversing that decision, but that should happen on another ticket.